runtime/include/QuantumDevice.hpp

```diff
@@ -55,310 +57,558 @@ struct QuantumDevice {
+     *
+     * See also `Observable` for observable semantics.
+     *
+     * @param obsKey The ID of the constructed observable.
+     *
+     * @return `double` The variance of the observable.
      */
-    virtual void Probs(DataView<double, 1> &probs) = 0;
+    virtual auto Var(ObsIdType obsKey) -> double { RT_FAIL("Var is unsupported by device"); }
 
     /**
-     * @brief Compute the probabilities for a subset of the full system.
+     * @brief (Optional) Get the full quantum state of all qubits.
+     *
+     * Typically for devices with statevector simulation capabilities.
+     *
+     * The output of this operation is the quantum statevector in the computational basis
+     * on all currently allocated qubits. When possible, the result should be produced without
+     * affecting the internal quantum state.
      *
-     * @param probs The pre-allocated `DataView<double, 1>`
-     * @param wires Wires will restrict probabilities to a subset of the full system
+     * The result of this operation must be written into the `state` argument buffer.
+     *
+     * @param state Pre-allocated buffer for the quantum state.
      */
-    virtual void PartialProbs(DataView<double, 1> &probs,
-                              const std::vector<QubitIdType> &wires) = 0;
+    virtual void State([[maybe_unused]] DataView<std::complex<double>, 1> &state)
+    {
+        RT_FAIL("State is unsupported by device");
+    }
+
+    // ----------------------------------------
+    //  QUANTUM DERIVATIVES
+    // ----------------------------------------
 
     /**
-     * @brief Compute samples with the number of shots on the entire wires,
-     * returing raw samples.
+     * @brief (Optional) Compute the gradient/Jacobian of a quantum circuit.
+     *
+     * For devices with internal differentiation capabilities (e.g. simulator with reverse-mode AD).
+     *
+     * Performing differentiation inside a device can be advantageous for performance compared to
+     * device-agnostic methods like the Parameter-Shift technique. If supported, Catalyst will
+     * communicate the start and end of the "forward pass" via the `StartTapeRecording` and
+     * `StopTapeRecording` functions. The `Gradient` call then requests the "reverse pass" with
+     * respect to the recorded circuit and observables. A sample implementation of circuit
+     * recording is available in `CacheManager.hpp`.
+     *
+     * The output is an array of gradients (i.e. the Jacobian), one for each expectation value /
+     * observable in the recorded program. The length of the gradient is determined by the number
+     * of gate parameters in the recorded circuit, or by the optional `trainParams` argument. If
+     * provided, `trainParams` can customize which parameter values participate in differentiation.
+     *
+     * The result of this operation must be written into the `gradients` argument buffer.
+     *
+     * @param gradients The vector of pre-allocated `DataView<double, 1>*`
```

Comment:
just to highlight that we store the flatten format of gradients:
```suggestion
     * @param gradients The flatten vector of pre-allocated `DataView<double, 1>*`
```

Reply:
https://github.com/PennyLaneAI/catalyst/pull/1680/commits/e7c92a7ead42d8ac4d000addedfa0cfa5c69405f